core/server/master/src/main/java/alluxio/master/file/ListStatusPartial.java

```
@@ -0,0 +1,208 @@
+    return partialPath;
+  }
+
+  /**
+   * If listing using a prefix and a partial path, this will compute whether the prefix
+   * exits in the partial path.
+   *
+   * @param partialPath the components of the path from where to start the listing
+   * @return the components of the prefix split by the delimiter /
+   * @throws InvalidPathException if the path in prefixComponents does not exist in partialPath
+   */
+  static List<String> checkPrefixListingPaths(
+      ListStatusContext context, List<String> partialPath)
+      throws InvalidPathException {
+    Optional<ListStatusPartialPOptions.Builder> pOptions = context.getPartialOptions();
+    if (!pOptions.isPresent()) {
+      return Collections.emptyList();
+    }
+    List<String> prefixComponents;
+    ListStatusPartialPOptions.Builder partialOptions = pOptions.get();
+    if (!partialOptions.getPrefix().isEmpty()) {
+      // compute the prefix as path components, removing the first empty string
+      prefixComponents = Arrays.stream(PathUtils.getPathComponents(
+              new AlluxioURI(
+                  partialOptions.getPrefix()).getPath())).skip(1)
+          .collect(Collectors.toList());
+    } else {
+      prefixComponents = Collections.emptyList();
+    }
+    // we only have to check the prefix if we are doing a partial listing,
+    // and we are not on the initial partial listing call
+    if (partialPath.isEmpty()
+        || !(partialOptions.hasOffset() && partialOptions.getOffset() != 0)) {
+      return prefixComponents;
+    }
+    // for each component the prefix must be the same as the partial path component
+    // except at the last component where the prefix must be contained in the partial path component
+    if (!PathUtils.hasPrefixComponentsCanBeLonger(partialPath, prefixComponents)) {
+      throw new InvalidPathException(
+          ExceptionMessage.PREFIX_DOES_NOT_MATCH_PATH
+              .getMessage(prefixComponents, partialPath));
+    }
```

move the check up to fail fast?

---

It is done here because the path and prefix components are generated and checked individually first.